Preliminary search of AnVIL repos using github api

avahoffman opened the pull request:
### Purpose/implementation Section

#### What changes are being implemented in this Pull Request?

These changes use github actions to search the API and pull out any repos that have the word "AnVIL" in the title and belong to the jhudsl organization. The target for the request is https://api.github.com/search/repositories?q=AnVIL+user:jhudsl .

Currently, loading all the libraries necessary in the line below takes quite a long time (5+ minutes). Would love to cut down on this if possible.
```
sudo Rscript -e "install.packages(c('optparse', 'curl', 'httr', 'jsonlite', 'dplyr', 'readr'))"
```

The list of repos becomes an artifact (`resources/AnVIL_repos.tsv`) and I've had it read in to the index page to look like this:
<img width="980" alt="image" src="https://user-images.githubusercontent.com/15618412/157131450-f68810ca-b0c0-4253-8de7-e12dc702ac4f.png">

avahoffman commented:
@KatherineCox I have a little something working here! It should hopefully be able to grow with the addition of new AnVIL repos. Would love to hear your thoughts!

avahoffman commented on .github/workflows/pull_request.yml:
```diff
@@ -57,6 +57,53 @@ jobs:
+          GH_PAT: ${{ secrets.GH_PAT }}
+        run: |
+          sudo apt-get install libcurl4-openssl-dev
+          sudo Rscript -e "install.packages(c('optparse', 'curl', 'httr', 'jsonlite', 'dplyr', 'readr'))"
```
This takes forever

KatherineCox replied:
Is that something we could make a custom Docker image to solve?

KatherineCox replied:
Also, in general (and especially if it's slow), do we want it to run for every PR?  We could make it a manual trigger, or, if there's a way to set GHA to watch a specific file, maybe we could have it run whenever sync.yml is changed (assuming that a change to sync.yml is often because a new "child" repo has been added)?

avahoffman replied:
Added the container. Super fast now :)

KatherineCox commented:
Overall:
- This is awesome!  Love it!
- Nice job making it both possible and straightforward to exclude specific repos

Cosmetic suggestion:
- Could we sub underscores with spaces when printing out book names?  That seems like it would read better and generally not cause problems
- More fragile would be to replace `_Book_` with `:`, so we'd get titles like "AnVIL: Getting Started".  Prettier, but more breakable since it relies on consistent naming of the repos.

avahoffman commented:
> Overall:
> 
>     * This is awesome!  Love it!
> 
>     * Nice job making it both possible and straightforward to exclude specific repos
> 
> 
> Cosmetic suggestion:
> 
>     * Could we sub underscores with spaces when printing out book names?  That seems like it would read better and generally not cause problems
> 
>     * More fragile would be to replace `_Book_` with `:`, so we'd get titles like "AnVIL: Getting Started".  Prettier, but more breakable since it relies on consistent naming of the repos.

Great suggestions. The underscore is easy. Might be able to strip AnVIL_Book altogether.. Let me play around with it!

avahoffman commented:
> Are you thinking this would run only in this repo, and then sync the results with downstream repos? Or that the downstream repos would run it as part of their rendering process?

I was thinking all of the repos would run this action, much like the spell or url check.. But maybe it makes more sense to sync. Do you have a strong preference? Sync would mean periodically rerunning the action manually in this repo, versus having it happen in the background as part of the pull requests. Might also be a little annoying having PRs to merge in the downstream repos..

KatherineCox commented:
> I was thinking all of the repos would run this action, much like the spell or url check.. But maybe it makes more sense to sync. Do you have a strong preference? Sync would mean periodically rerunning the action manually in this repo, versus having it happen in the background as part of the pull requests. Might also be a little annoying having PRs to merge in the downstream repos..

hrmmm, I was originally thinking syncing, but that was before we had a handy-dandy robot that *could* do it automatigically for each repo

My only strong opinion is that I don't want to make PR rendering take much longer.  If I'm making smaller edits to a book, these days I usually don't bother rendering locally, but then I have to wait through the GitHub render to realize I messed up the spacing on my numbering or forgot to make my slides public or whatever small problems show up.  So I don't really want to extend that wait if I can avoid it.

But if we can mitigate that (Docker image?) or have an option to turn it off, then I don't think I have a strong opinion.

**If done centrally**:

Pros:

- could act as a sort of "release" mechanism for books - If there's an AnVIL book under development but not ready to be broadly advertised, then it can be excluded in the template repository.  Then, as part of "releasing" v 1.0 of a book, it can be un-excluded from the template and it'll get passed along to children.

Cons:

- potential complications with PRs
- need to make sure the template repo stays up-to-date

**If done for each repo**:

Pros:

- authors could customize which books show up, if there's a subset that makes most sense for the readers of their book
- authors could decide where to put the code for adding the table
- don't have to deal with merging PRs

Cons:

- list can get out of date if the repo isn't under active development.  Not really a big deal, we're not exactly cranking out piles of books :)


...that's all I'm coming up with for now.  Looking at my list, I think I'd come down on the side of running it in the individual repos (mostly because it gives authors more flexibility) as long as we can minimize the time it adds to a PR render.

**Question**: Would this make it impossible for a book to render locally?  If so, that might be a reason to keep it central.

avahoffman commented:
> My only strong opinion is that I don't want to make PR rendering take much longer. If I'm making smaller edits to a book, these days I usually don't bother rendering locally, but then I have to wait through the GitHub render to realize I messed up the spacing on my numbering or forgot to make my slides public or whatever small problems show up. So I don't really want to extend that wait if I can avoid it.

I added the container into this specific action. Takes almost no time, so good there :)

> ...that's all I'm coming up with for now. Looking at my list, I think I'd come down on the side of running it in the individual repos (mostly because it gives authors more flexibility) as long as we can minimize the time it adds to a PR render.

Same - this is the solution I've added (for now).

> **Question**: Would this make it impossible for a book to render locally? If so, that might be a reason to keep it central.

Yes. I'll add a try loop so it doesn't error out locally.

KatherineCox commented:
> I added the container into this specific action. Takes almost no time, so good there :)

Awesome!

> Yes. I'll add a try loop so it doesn't error out locally.

Sounds good. 👍

avahoffman commented:
See #52 for the local rendering trycatch
